File: agent/internal/tool/image_search.go
package tool

import (
	"context"
	"fmt"

	"github.com/xz-ephyr/raw-code/agent/pkg/api"
)

func imageSearchTool() ToolDef {
	return ToolDef{
		Definition: api.ToolDefinition{
			Name:        "image_search",
			Description: "Search for images on the web by query. Returns image URLs, dimensions, and source pages.",
			Category:    "research",
			Parameters: map[string]api.ParamDef{
				"query":       {Type: "string", Description: "The image search query", Required: true},
				"max_results": {Type: "number", Description: "Max images (1-20)", Required: false, Default: 5},
				"safe_search": {Type: "boolean", Description: "Enable safe search filtering", Required: false, Default: true},
			},
		},
		Handler: func(ctx context.Context, e *Executor, params map[string]any) (any, error) {
			query, _ := params["query"].(string)
			if query == "" {
				return nil, fmtError("query is required")
			}
			maxResults, _ := params["max_results"].(float64)
			if maxResults == 0 {
				maxResults = 5
			}
			safeSearch := true
			if v, ok := params["safe_search"].(bool); ok {
				safeSearch = v
			}

			return postToExpress(ctx, e.expressURL, "imageSearch", map[string]any{
				"query":      query,
				"maxResults": int(maxResults),
				"safeSearch": safeSearch,
			})
		},
	}
}

File: src/services/tools/web/imageSearch.ts
import { z } from 'zod';
import type { ToolDef } from '../types';
import { WebSearchService } from '../../WebSearchService';

export const imageSearchTool: ToolDef = {
  name: 'image_search',
  description: 'Search the web for images. Returns image URLs, titles, and source pages.',
  category: 'web',
  inputSchema: z.object({
    query: z.string().describe('The image search query. Descriptive terms work best.'),
    maxResults: z.number().min(1).max(5).optional().default(5).describe('Number of image results (1–5).'),
    safeSearch: z.boolean().optional().default(true).describe('Filter out explicit content.'),
  }),
  execute: async ({ query, maxResults, safeSearch }) => {
    try {
      const result = await WebSearchService.imageSearch({ query, maxResults: Math.min(maxResults ?? 5, 5), safeSearch });
      return {
        results: result.results.map(r => ({
          title: r.title, imageUrl: r.imageUrl, sourceUrl: r.sourceUrl,
          width: r.width, height: r.height,
        })),
        totalResults: result.totalResults,
      };
    } catch (error: any) {
      throw new Error(error.message || 'Image search failed.');
    }
  },
};
